encoder.go

```diff
@@ -102,6 +98,20 @@ func (e *encoder) setFieldByType(current reflect.Value, namespace []byte, idx in
 			return
 		}
 	}
+	{
+		if t := v.Type(); t.Kind() == reflect.Ptr && v.IsNil() {
+			return
+		}
```

Comment:
Is this check necessary?

1. Below this logic `reflect.Ptr` is handled and returned.
2. What is someone wanted to set a default value when it was nil from the custom marshaller, would this prevent that?

Reply:
So I've played around with this, and could get something sorted that didn't risk a nil pointer in some form.

For example, by removing this check and making one or two other modifications, I ended up allowing this following to work:
```go
func (c *custom) MarshalForm() ([]string, error) {
    if c == nil {
        return // some default value
    }

    return // real handling
}

func main() {
    var c *custom

    NewEncoder().Encode(c)
}
```

However, if `MarshalForm` is changed to have a value receiver, then we'd end up with a nil panic because the nil `*custom` passes the `form.Marshaler` implementation check, then have its `MarshalForm` function called, causing the panic once dereferenced to the receiver.

All this playing about ended up having me take a read of the `json.Marshaler` implementation in the standard library. It's behaviour is basically just, if the item being marshalled is nil, write `"null"` (see [here](https://github.com/golang/go/blob/07fc59199b9522bfe0d14f35c4391394efc336c9/src/encoding/json/encode.go#L431-L434), and [here](https://github.com/golang/go/blob/07fc59199b9522bfe0d14f35c4391394efc336c9/src/encoding/json/encode.go#L453-L457)).

The way the std lib handles its reflection checks is much better than what I've done here so I'm gonna do a small rewrite, but do let me know what behaviour you would prefer regarding this.